File: 백준/Silver/2579. 계단 오르기/계단 오르기.swift
import Foundation

let n = Int(readLine()!)!
var stairs = [0] // 0번 인덱스를 비워두고 시작

for _ in 1...n {
    stairs.append(Int(readLine()!)!)
}

var dp = [Int](repeating: 0, count: n + 1)

dp[1] = stairs[1]
if n >= 2 {
    dp[2] = stairs[1] + stairs[2]
}

if n >= 3 {
    for i in 3...n {
        dp[i] = max(dp[i-2] + stairs[i], dp[i-3] + stairs[i-1] + stairs[i])
    }
}

print(dp[n])

File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
import sys

#n, n-1, n-2, n-3, n-4, n-5, 시작

input = sys.stdin.readline

n = int(input())

stairs = [0] * 301
for index in range(1, n + 1):
    stairs[index] = int(input())
    
dp = [0] * 301
dp[1] = stairs[1]
dp[2] = stairs[1] + stairs[2]
dp[3] = max(stairs[1] + stairs[3], stairs[2] + stairs[3])

#점화식 계산
for index in range(4, n + 1):#n-1 밟는 경우, n-2 밟는 경우
    dp[index] = max(stairs[index] + stairs[index-1] + dp[index-3], stairs[index] + dp[index-2])

print(dp[n])
